Sources_v3/Workers/Background/MessageSender.swift

```diff
@@ -117,70 +117,70 @@ private class MessageSendingQueue<ExtraData: ExtraDataTypes> {
         let wasEmpty = self.requests.isEmpty
         requests.forEach { self.requests.insert($0) }
         if wasEmpty {
-            dispatchQueue.async { [weak self] in
-                self?.sendNextMessage()
-            }
+            sendNextMessage()
         }
     }
     
     /// Gets the oldes message from the qeue and tries to send it.
     private func sendNextMessage() {
-        // Sort the messages and send the oldest one
-        guard let request = requests.sorted(by: { $0.createdLocallyAt < $1.createdLocallyAt }).first else { return }
-        
-        // Check the message with the given id is still in the DB.
-        guard let dto = database.backgroundReadOnlyContext.message(id: request.messageId) else {
-            log.error("Trying to send a message with id \(request.messageId) but the message was deleted.")
-            removeRequestAndContinue(request)
-            return
-        }
-        
-        // Check the message still have `pendingSend` state.
-        guard dto.localMessageState == .pendingSend else {
-            log.info("Skipping sending message with id \(dto.id) because it doesn't have `pendingSend` local state.")
-            removeRequestAndContinue(request)
-            return
-        }
-        
-        let cid = try! ChannelId(cid: dto.channel.cid)
-        let requestBody = dto.asRequestBody() as MessageRequestBody<ExtraData>
-        
-        // Change the message state to `.sending` and the proceed with the actual sending
-        database.write({
-            let messageDTO = $0.message(id: request.messageId)
-            messageDTO?.localMessageState = .sending
+        dispatchQueue.async { [weak self] in
+            // Sort the messages and send the oldest one
+            guard let request = self?.requests.sorted(by: { $0.createdLocallyAt < $1.createdLocallyAt }).first else { return }
             
-        }, completion: { [weak self] error in
-            if let error = error {
-                log.error("Error changin localMessageState message with id \(request.messageId) to `sending`: \(error)")
-                self?.markMessageAsFailedToSend(id: request.messageId) {
-                    self?.removeRequestAndContinue(request)
-                }
-                
+            // Check the message with the given id is still in the DB.
+            guard let dto = self?.database.backgroundReadOnlyContext.message(id: request.messageId) else {
+                log.error("Trying to send a message with id \(request.messageId) but the message was deleted.")
+                self?.removeRequestAndContinue(request)
                 return
             }
```

Comment:
The DTO must be obtained inside the `backgroundReadOnlyContext.peform { }` otherwise you can't safely retain it and work with it lately. It can be modified in the meantime from another queue.

Reply:
yes, but this is in the scope of `CIS-333`. The motivation for fixing it in `MessageEditor` is here https://github.com/GetStream/stream-chat-swift/pull/487#discussion_r486959623

Reply:
Right, sorry for the noise :)